research/delf/delf/python/training/model/global_model.py

```diff
@@ -0,0 +1,290 @@
+  'VGG19': 'http://cmp.felk.cvut.cz/cnnimageretrieval_tf'
+           '/SFM120k_VGG19_gem_learned_whitening_config.pkl'
+}
+
+# Possible global pooling layers.
+_POOLING = {
+  'mac': pooling_layers.MAC,
+  'spoc': pooling_layers.SPoC,
+  'gem': pooling_layers.GeM
+}
+
+# Output dimensionality for supported architectures.
+_OUTPUT_DIM = {
+  'VGG16': 512,
+  'VGG19': 512,
+  'ResNet50': 2048,
+  'ResNet101': 2048,
+  'ResNet101V2': 2048,
+  'ResNet152': 2048,
+  'DenseNet121': 1024,
+  'DenseNet169': 1664,
+  'DenseNet201': 1920,
+  'EfficientNetB5': 2048,
+  'EfficientNetB7': 2560
+}
+
+
+class GlobalFeatureNet(tf.keras.Model):
+  """Instantiates global model for image retrieval.
+
+  This class implements the [GlobalFeatureNet](
+  https://arxiv.org/abs/1711.02512) for image retrieval. The model uses a
+  user-defined model as a backbone.
+  """
+
+  def __init__(self, architecture='ResNet101', pooling='gem',
+               whitening=False, pretrained=True):
+    """GlobalFeatureNet network initialization.
+
+    Args:
+      architecture: Network backbone 'VGG16'/'VGG19'/'ResNet50'/'ResNet101/
+        'ResNet101V2'/'ResNet152'/'DenseNet121'/'DenseNet169'/'DenseNet201'.
+      pooling: Pooling method used 'mac'/'spoc'/'gem'.
+      whitening: Bool, whether to use whitening.
+      pretrained: Bool, whether to initialize the network with pretrained
+        weights.
+    """
+    super(GlobalFeatureNet, self).__init__()
+
+    # Get standard output dimensionality size.
+    dim = _OUTPUT_DIM[architecture]
+
+    if pretrained:
+      # Initialize with network pretrained on imagenet.
+      net_in = getattr(tf.keras.applications, architecture)(include_top=False,
+                                                            weights="imagenet")
+    else:
+      # Initialize with random weights.
+      net_in = getattr(tf.keras.applications, architecture)(include_top=False,
+                                                            weights=None)
+
+    # Initialize `feature_extractor`. Take only convolutions for
+    # `feature_extractor`, always end with ReLU to make last activations
+    # non-negative.
+    if architecture.lower().startswith('densenet'):
+      net_in.add(tf.keras.layers.ReLU())
+
+    # Initialize pooling.
+    pool = _POOLING[pooling]()
+
+    # Initialize whitening.
+    if whitening:
+      if pretrained and architecture in _WHITENING_CONFIG:
+        # If precomputed whitening for the architecture exists,
+        # the fully-connected layer is going to be initialized according to
+        # the precomputed layer configuration.
+        global_features_utils.debug_and_log(
+          ">> {}: for '{}' custom computed whitening '{}' is used."
+            .format(os.getcwd(), architecture,
+                    os.path.basename(_WHITENING_CONFIG[architecture])))
+        # The layer configuration is downloaded to the `data_root` folder.
+        whiten_dir = os.path.join(FLAGS.data_root, architecture)
```

Comment:
you are using some FLAGS which are not defined in the same module. This is generally discouraged, and the recommended pattern would be to add a new `data_root` argument in the init(). Can you please do this?